src/EOS.hpp

```diff
@@ -49,42 +59,111 @@ template <typename problem_t> class EOS
 
 template <typename problem_t>
 AMREX_FORCE_INLINE AMREX_GPU_HOST_DEVICE auto EOS<problem_t>::ComputeTgasFromEint(amrex::Real rho, amrex::Real Eint,
-										  std::optional<amrex::GpuArray<amrex::Real, nmscalars_>> massFractions)
+										  std::optional<amrex::GpuArray<amrex::Real, nmscalars_>> massScalars)
     -> amrex::Real
 {
 	// return temperature for an ideal gas
 	amrex::Real Tgas = NAN;
-	if constexpr (gamma_ != 1.0) {
-		const amrex::Real c_v = boltzmann_constant_ / (mean_molecular_weight_ * (gamma_ - 1.0));
-		Tgas = Eint / (rho * c_v);
+
+	if constexpr (PRIMORDIAL_CHEM_ENABLED) {
+		burn_t chemstate;
+		chemstate.rho = rho;
+		chemstate.e = Eint / rho;
+		// initialize array of number densities
+		for (int ii = 0; ii < NumSpec; ++ii) {
+			chemstate.xn[ii] = -1.0;
+		}
+
+		if (massScalars.has_value()) {
+			const auto &massArray = massScalars.value();
+			for (int nn = 0; nn < nmscalars_; ++nn) {
+				chemstate.xn[nn] = massArray[nn] / spmasses[nn]; // massScalars are partial densities (massFractions * rho)
+			}
+		}
+
+		eos(eos_input_re, chemstate);
+		amrex::Real Tgas = chemstate.T;
```

Comment:
This `Tgas` shadows the `Tgas` in the enclosing scope. So the values of Tgas returned will always be `NAN`. You can fix this by just changing this line to `Tgas = chemstate.T;`.